### PR TITLE
DynamicSubsetFormula

### LINKED ISSUE #665: Dynamic `subsetFormula`
Currently `subsetFormula` let you limit visible rows to the user. Problem is that this expression is static and defined once inside configuration file.

Inspired by @cedrics in #501, we're thinking about something like `dynamicSubset`.

This is a javascript function defined in configuration file. It will receive request data (we're thinking about headers) and should return valid plywood expression that could be used as filter argument. 

Such function would be called at every request and resulting expression would be added with `and` operator to filter on `$main`. 

Feedback is welcome, and WIP pull request would be attached in few days.

### PR DESCRIPTION
**Pull request by adrianmroz**
Closes #665

### CODE REVIEW

File: src/server/routes/plywood/plywood.ts [hunk 1]
```diff
@@ -17,16 +17,28 @@
...
 import { Request, Response, Router } from "express";
-import { Dataset, Expression } from "plywood";
+import { IncomingHttpHeaders } from "http";
+import { Dataset, Expression, RefExpression } from "plywood";
+import { DynamicSubsetFormula } from "../../../common/models/dynamic-subset-formula/dynamic-subset-formula";
 import { checkAccess } from "../../utils/datacube-guard/datacube-guard";
 import { SettingsGetter } from "../../utils/settings-manager/settings-manager";
 
+function applySubset(expression: Expression, dynamicSubsetFormula: DynamicSubsetFormula, headers: IncomingHttpHeaders): Expression {
+  const subsetFilter = dynamicSubsetFormula.fn(headers);
```
Comment (adrianmroz) [inline src/server/routes/plywood/plywood.ts:27]: Add some checks:

- does `dynamicSubsetFormula` exists?
- does it throw?
- is returned value a valid Expression?

File: src/common/models/dynamic-subset-formula/dynamic-subset-formula.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IncomingHttpHeaders } from "http";
+import { Instance } from "immutable-class";
+import { Expression } from "plywood";
+import { Unary } from "../../utils/functional/functional";
+
+export type DynamicSubsetFormulaFn = Unary<IncomingHttpHeaders, Expression>;
```
Comment (adrianmroz) [inline src/common/models/dynamic-subset-formula/dynamic-subset-formula.ts:22]: What values could we pass there? Full request object? Only headers? Maybe more fields on request object like subdomains, path, ip etc?

File: src/server/routes/plywood/plywood.ts (hunk 1 above)
Comment (adrianmroz) [reply src/server/routes/plywood/plywood.ts]: done

File: src/common/models/dynamic-subset-formula/dynamic-subset-formula.ts (hunk 2 above)
Comment (mkuthan) [reply src/common/models/dynamic-subset-formula/dynamic-subset-formula.ts:22]: let's start from headers and make design extensible

File: src/server/routes/plywood/plywood.ts [hunk 3]
```diff
@@ -17,16 +17,44 @@
...
-import { Dataset, Expression } from "plywood";
+import { IncomingHttpHeaders } from "http";
+import { Dataset, Expression, RefExpression } from "plywood";
+import { isFunction } from "util";
+import { DynamicSubsetFormula } from "../../../common/models/dynamic-subset-formula/dynamic-subset-formula";
+import { isNil } from "../../../common/utils/general/general";
 import { checkAccess } from "../../utils/datacube-guard/datacube-guard";
 import { SettingsGetter } from "../../utils/settings-manager/settings-manager";
 
+function filterMain(expression: Expression, filter: Expression): Expression {
```
Comment (mkuthan) [inline src/server/routes/plywood/plywood.ts:28]: I would move filterMain and applySubset somewhere. Route is not the place for this kind of logic.

File: src/server/routes/plywood/plywood.ts [hunk 4]
```diff
@@ -49,9 +77,9 @@ export function plywoodRouter(getSettings: SettingsGetter) {
       }
     }
 
-    let ex: Expression = null;
+    let parsedExpression: Expression = null;
```
Comment (mkuthan) [inline src/server/routes/plywood/plywood.ts:80]: Encapsulate parsing somewhere (perhaps together with applySubset / filterMain)

Comment (adrianmroz) [reply src/server/routes/plywood/plywood.ts:80]: But parsing here is simple, complex logic is around catching error and sending appropriate response which I think is role of route.

File: src/server/routes/plywood/plywood.ts (hunk 3 above)
Comment (adrianmroz) [reply src/server/routes/plywood/plywood.ts:28]: Of course!

File: src/common/models/dynamic-subset-formula/dynamic-subset-formula.ts (hunk 2 above)
Comment (adrianmroz) [reply src/common/models/dynamic-subset-formula/dynamic-subset-formula.ts:22]: We need to pass plywood dependency, so function can use it's DSL and return real Expression

**Review by fpoon — APPROVED** [review]

**Review by piter75 — APPROVED** [review]

**Pull request merged**